torch/csrc/dynamo/guards.cpp

```diff
@@ -3681,6 +3969,52 @@ void install_no_tensor_aliasing_guard(
   }
 }
 
+void install_storage_overlapping_guard_with_checker(
+    std::shared_ptr<StorageOverlapChecker> checker,
+    const py::list& guard_managers,
+    py::object verbose_code_parts,
+    bool overlapping) {
+  if (guard_managers.size() == 0) {
+    // If there are no GuardManagers, there's no need to create a
+    // STORAGE_OVERLAPPING guard.
+    return;
+  }
+
+  std::shared_ptr<RelationalGuard> guard =
+      std::make_shared<STORAGE_OVERLAPPING>(
+          overlapping, checker, verbose_code_parts);
+  py::cast<GuardManager*>(guard_managers[0])
+      ->get_root()
+      ->add_relational_guard_resetter(guard);
+  for (const auto& guard_manager : guard_managers) {
+    py::cast<GuardManager*>(guard_manager)->add_leaf_guard(guard);
+  }
+}
+
+void install_storage_overlapping_guard(
```

Comment:
cc @anijain2305 for review too. There's more context on what these guards are in the PR description, but the TLDR is:

(1) they are needed in the case where we have `f(x, y, z)`, one of our inputs is mutated, and x/y/z share storages. We need to carefully check the sizes/strides/offsets of each tensor to see if they **truly** alias, or if they don't actually overlap in memory, since this affects how we functionalize. These guards are only actually necessary to use if you have a graph with input mutation, input aliasing, and dynamic shapes

(2) executing the guard in pure python is slow, so Yukio is moving it to C++